### PR TITLE
Process GitHub Push Hook notifications

### PR DESCRIPTION
**Pull request by reset**
This commit changes how we make authenticated requests to the GitHub API
on behalf of the user and adds the ability to process GitHub Webhook
notifications for Ping and Push.

In order to support Ping and Push notifications for Private repositories
we needed to switch to integrating with GitHub using a full GitHub App
instead of just an OAuth App. This PR changes the github-api-client to
only make requests which a GitHub App could make and signs the commits
appropriately.

The authorization process has changed to a per repo basis and no longer
requires a user to setup a personal access token.

tenor-169787551

### CODE REVIEW

**Comment by thesentinels** [comment]
Thanks for the pull request! Here is what will happen next:
 1. Your PR will be reviewed by the maintainers
 2. If everything looks good, one of them will `approve` it, and your PR will be merged.

Thank you for contributing!

File: components/builder-api/src/github.rs [hunk 1]
```diff
@@ -0,0 +1,222 @@
...
+        _ => return Ok(Response::with(status::BadRequest)),
+    };
+    match event {
+        GitHubEvent::Ping => Ok(Response::with(status::Ok)),
+        GitHubEvent::Push => handle_push(req),
+    }
+}
+
+fn handle_push(req: &mut Request) -> IronResult<Response> {
+    // JW TODO: THIS NEEDS TO BE AUTHENTICATED
```
Comment (reset) [inline components/builder-api/src/github.rs]: I'm going to handle decrypting the private key before we push this to live

File: components/builder-api/src/github.rs [hunk 2]
```diff
@@ -0,0 +1,222 @@
...
+        plans.retain(|plan| match cfg.get(&plan.name) {
+            Some(project) => hook.changed().iter().any(|f| project.triggered_by(f)),
+            None => false,
+        })
+    }
+    build_plans(req, plans)
+}
+
+fn build_plans(req: &mut Request, plans: Vec<Plan>) -> IronResult<Response> {
+    // JW TODO: Validate that this repository is where these plans belong. You could theoretically
```
Comment (reset) [inline components/builder-api/src/github.rs]: I'll also handle this before we push to live

File: components/builder-protocol/src/originsrv.rs [hunk 3]
```diff
@@ -980,6 +980,38 @@ impl Serialize for OriginProject {
     }
 }
 
+// impl Routable for ProjectStateSet {
```
Comment (christophermaier) [inline components/builder-protocol/src/originsrv.rs]: Did you mean to add this?

File: components/github-api-client/src/types.rs [hunk 4]
```diff
@@ -0,0 +1,392 @@
...
+#[derive(Debug, Default, Clone, Serialize, Deserialize)]
+#[serde(default)]
+pub struct GitHubAuthor {
+    /// Public name of commit author
+    pub name: String,
+    /// Public email of commit author
+    pub email: String,
+    /// Display name of commit author
+    pub username: String,
+}
```
Comment (christophermaier) [inline components/github-api-client/src/types.rs]: Are the values of these just `""` if the user doesn't have them set? Should these be `Option<String>` (at least, `email` and `name`)? Does it matter much for our usecase?

(Ditto for the other types in this module that are `String` instead of `Option<String>`.)

File: components/github-api-client/src/types.rs [hunk 5]
```diff
@@ -0,0 +1,392 @@
...
+pub struct SearchItem {
+    pub name: String,
+    pub path: String,
+    pub sha: String,
+    pub url: String,
+    pub git_url: String,
+    pub html_url: String,
+    pub repository: Repository,
+    pub score: f64,
+}
```
Comment (christophermaier) [inline components/github-api-client/src/types.rs]: I'm amazed there's not a library for a Github client in Rust yet!

File: components/builder-admin/src/server/mod.rs [hunk 6]
```diff
@@ -35,7 +35,7 @@ impl HttpGateway for AdminSrv {
 
     fn add_middleware(config: Arc<Self::Config>, chain: &mut iron::Chain) {
         chain.link(persistent::Read::<GitHubCli>::both(
-            GitHubClient::new(&*config),
+            GitHubClient::new((*config).github.clone()),
```
Comment (christophermaier) [inline components/builder-admin/src/server/mod.rs]: Rust question time: What's the significance of the parentheses around `*config`?

File: components/builder-protocol/src/originsrv.rs (hunk 3 above)
Comment (reset) [reply components/builder-protocol/src/originsrv.rs]: No, it was left over from a really old branch. Thanks!

File: components/github-api-client/src/types.rs (hunk 4 above)
Comment (reset) [reply components/github-api-client/src/types.rs]: When we use `Option<String>` it's when the API can send `null` for a field. I should probably remove `serde(default)` from these structs which is how you'll end up with an empty string instead of a runtime error

File: components/builder-admin/src/server/mod.rs (hunk 6 above)
Comment (reset) [reply components/builder-admin/src/server/mod.rs]: Actually now that I go to type this out it might be overkill. I was dereferencing the `Arc` to get at the inner value but `.github` will do the same thing

File: components/github-api-client/src/types.rs (hunk 5 above)
Comment (reset) [reply components/github-api-client/src/types.rs]: There are a few but nothing with enough coverage. The good thing is that we are only going to be using a portion of the GitHub API so we can create a well tested sub set and not need to rely on an upstream dependency

Comment (christophermaier) [reply components/github-api-client/src/types.rs]: Admittedly, that _is_ pretty nice. 👍

File: components/builder-admin/src/server/mod.rs (hunk 6 above)
Comment (christophermaier) [reply components/builder-admin/src/server/mod.rs]: Cool.

**Comment by christophermaier** [comment]
tenor-58292932

Looks good; feel free to merge whenever you like.

**Comment by reset** [comment]
@thesentinels approve

**Comment by thesentinels** [comment]
:metal: I am testing your branch against master before merging it. We do this to ensure that the master branch is never failing tests.

**Comment by chefsalim** [comment]
tenor-146797911

**Comment by thesentinels** [comment]
:neckbeard: Travis CI has [started testing this PR](https://travis-ci.org/habitat-sh/habitat/builds/282354706).

File: components/builder-protocol/protocols/originsrv.proto [hunk 7]
```diff
@@ -341,7 +341,7 @@ message OriginProject {
   optional uint64 owner_id = 7;
   optional string vcs_type = 8;
   optional string vcs_data = 9;
-  optional uint64 vcs_installation_id = 12;
+  optional uint32 vcs_installation_id = 12;
```
Comment (mwrock) [inline components/builder-protocol/protocols/originsrv.proto]: I think a 32 bit is plenty but curious why the change here since it seems our convention is 64.

**Comment by thesentinels** [comment]
:sparkling_heart: Travis CI [reports this PR passed](https://travis-ci.org/habitat-sh/habitat/builds/282354706).

It always makes me feel nice when humans approve of one anothers work. I'm merging this PR now.

I just want you and the contributor to answer me one question:

gif-keyboard-3280869874741411265

File: components/builder-protocol/protocols/originsrv.proto (hunk 7 above)
Comment (reset) [reply components/builder-protocol/protocols/originsrv.proto]: Our convention is 64 but their API is sending 32

File: components/github-api-client/src/client.rs [hunk 8]
```diff
@@ -27,42 +24,66 @@ use hyper::net::HttpsConnector;
...
-            web_url: config.github_web_url().to_string(),
-            client_id: config.github_client_id().to_string(),
-            client_secret: config.github_client_secret().to_string(),
-            app_id: config.github_app_id(),
-            app_private_key_path: config.github_app_private_key_path(),
+            url: config.url,
+            web_url: config.web_url,
+            client_id: config.client_id,
+            client_secret: config.client_secret,
+            app_private_key: config.app_private_key,
```
Comment (mwrock) [inline components/github-api-client/src/client.rs]: did we intend to hard code the app id?

File: components/github-api-client/src/client.rs [hunk 9]
```diff
@@ -258,286 +259,94 @@ impl GitHubClient {
...
+where
+    T: ToString,
+{
+    let mut payload = jwt::Payload::new();
+    let header = jwt::Header::new(jwt::Algorithm::RS256);
+    let now = SystemTime::now().duration_since(UNIX_EPOCH).unwrap();
+    let expiration = now + Duration::from_secs(10 * 60);
+    payload.insert("iat".to_string(), now.as_secs().to_string());
+    payload.insert("exp".to_string(), expiration.as_secs().to_string());
+    payload.insert("iss".to_string(), 5565.to_string());
```
Comment (mwrock) [inline components/github-api-client/src/client.rs]: did we intend to hard code the app id?

File: components/builder-protocol/protocols/originsrv.proto (hunk 7 above)
Comment (mwrock) [reply components/builder-protocol/protocols/originsrv.proto]: Yeah I think that was my fault. Thats the type I used in our frank fork. Should be fine but in retrospect I'd have used u64.

File: components/github-api-client/src/client.rs (hunk 9 above)
Comment (reset) [reply components/github-api-client/src/client.rs]: Absolutely not, I've got this fixed in another branch ;)
